### DISCUSSION TITLE
🐞 Dagger call can't install dependency modules which refs that don't belong to a branch

### DISCUSSION THREAD

**Discussion by marcosnils** [dagger/dagger#12249] [Help]
### What is the issue?

While performing a `dagger call` in a module that has a dependency which points to a ref that doesn't exist on a branch in the target repo, the call will fail with the following message: 

Error: input: moduleSource.withContextDirectory.asModule resolve: failed to create module: select: failed t
o update module dependencies: failed to load module dependencies: select: failed to load pre-configured dependencies: failed to create module source from dependency: select: failed to resolve git src to commit: fai
led to checkout remote https://github.com/Dudesons/daggerverse: git error: exit status 128  

### Dagger version

v0.12.4

### Steps to reproduce

1 - Create a new Dagger module
2 - Call `dagger install github.com/Dudesons/daggerverse/secret-manager@v0.2.0`
3 - Use that dependency  in your module in any way. For example:

```go
func (m *Lele) Test() *dagger.Secret {
	return dag.SecretManager().Gcp().GetSecret("foo", "bar")
}
```
4. Clean up your existing Dagger engine so you start with a new one
5. Run `dagger call Test`  and you'll get the error displayed below

### Log output

Error: input: moduleSource.withContextDirectory.asModule resolve: failed to create module: select: failed t
o update module dependencies: failed to load module dependencies: select: failed to load pre-configured dependencies: failed to create module source from dependency: select: failed to resolve git src to commit: fai
led to checkout remote https://github.com/Dudesons/daggerverse: git error: exit status 128

**Comment by samalba** [comment]
Not sure what to think about this one. Even [Github complains about the commit that the tag resolves to](https://github.com/Dudesons/daggerverse/commit/5a0943df5ceca43ab24309e5331f0bac26208a58):

```
This commit does not belong to any branch on this repository, and may belong to a fork outside of the repository.
```

I guess the commit was tagged on a branch that has been deleted.

I agree that it's weird the mod install works but can't re-pull de dependency if the engine gets recreated...

**Comment by jpadams** [comment]
My error was slightly different on Mac

```
Full trace at https://dagger.cloud/dagger/traces/c652586e5eb6d7db7aaac9193fc24c94

✔ connect 3.3s
✘ initialize 1.2s
! input: moduleSource.withContextDirectory.asModule resolve: failed to create module: select: failed to update module dependencies: failed to load module dependencies: select: failed to load pre-configured dependencies: failed to create module source from dependency: select: failed to resolve git src to commit: failed to checkout remote https://github.com/Dudesons/daggerverse: git error: exit status 128
  ✔ resolving module ref 0.2s
  ✘ installing module 1.1s
  ! input: moduleSource.withContextDirectory.asModule resolve: failed to create module: select: failed to update module dependencies: failed to load module dependencies: select: failed to load pre-configured dependencies: failed to create module source from dependency: select: failed to resolve git src to commit: failed to checkout remote https://github.com/Dudesons/daggerverse: git error: exit status 128
    ✔ ModuleSource.resolveFromCaller: ModuleSource! 0.4s
    ✘ ModuleSource.asModule: Module! 0.6s
    ! failed to create module: select: failed to update module dependencies: failed to load module dependencies: select: failed to load pre-configured dependencies: failed to create module source from dependency: select: failed to resolve git src to commit: failed to checkout remote https://github.com/Dudesons/daggerverse: git error: exit status 128

Error: input: moduleSource.withContextDirectory.asModule resolve: failed to create module: select: failed to update module dependencies: failed to load module dependencies: select: failed to load pre-configured dependencies: failed to create module source from dependency: select: failed to resolve git src to commit: failed to checkout remote https://github.com/Dudesons/daggerverse: git error: exit status 128
stderr:
fatal: unable to read tree (5a0943df5ceca43ab24309e5331f0bac26208a58)
```

which seems odd, because we can do both
```
test ➤ dagger core git --url=https://github.com/Dudesons/daggerverse tag --name secret-manager/v0.2.0                     
Full trace at https://dagger.cloud/dagger/traces/6864fe5e144709dd3a00a2dc5a38c318

✔ connect 0.9s
✔ initialize 0.2s
✔ prepare 0.0s
✔ git(url: "https://github.com/Dudesons/daggerverse"): GitRepository! 0.0s
✔ GitRepository.tag(name: "secret-manager/v0.2.0"): GitRef! 0.0s
✔ GitRef.commit: String! 0.3s

_type: GitRef
commit: 5a0943df5ceca43ab24309e5331f0bac26208a58
test ➤ dagger core git --url=https://github.com/Dudesons/daggerverse commit --id 5a0943df5ceca43ab24309e5331f0bac26208a58
Full trace at https://dagger.cloud/dagger/traces/1d483ee2b966bf843ef2fbbafff492a7

✔ connect 0.8s
✔ initialize 0.2s
✔ prepare 0.0s
✔ git(url: "https://github.com/Dudesons/daggerverse"): GitRepository! 0.0s
✔ GitRepository.commit(id: "5a0943df5ceca43ab24309e5331f0bac26208a58"): GitRef! 0.0s
✔ GitRef.commit: String! 0.0s

_type: GitRef
commit: 5a0943df5ceca43ab24309e5331f0bac26208a58
```

**Comment by jedevc** [comment]
Aha, this is a more specific version of https://github.com/dagger/dagger/issues/8584.

Note that this is fiddly to reproduce because the buildkit internal git state can get into a very weird inconsistent state.
